### PR TITLE
MM-52797 : Drafts badge doesn't clear when parent message deleted and removed

### PR DESCRIPTION
**Pull request by M-ZubairAhmed**
#### Ticket Link
Jira https://mattermost.atlassian.net/browse/MM-52797

#### Screenshots
tba

#### Release Note

```release-note
Fixed an issue where the draft counter badge remained in case of a deleted parent post and its subsequent removal.
```

### CODE REVIEW

File: webapp/channels/src/components/threading/thread_viewer/index.ts [hunk 1]
```diff
@@ -78,7 +78,6 @@ function mapDispatchToProps(dispatch: Dispatch<GenericAction>) {
             getNewestPostThread,
             getPostThread,
             getThread: fetchThread,
-            removePost,
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/threading/thread_viewer/index.ts:81]: unused action

File: webapp/channels/src/components/threading/thread_viewer/thread_viewer.tsx [hunk 2]
```diff
@@ -40,7 +38,6 @@ export type Props = Attrs & {
         getNewestPostThread: (rootId: string) => Promise<any>|ActionFunc;
         getPostThread: (rootId: string, fetchThreads: boolean) => Promise<any>|ActionFunc;
         getThread: (userId: string, teamId: string, threadId: string, extended: boolean) => Promise<any>|ActionFunc;
-        removePost: (post: ExtendedPost) => void;
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/threading/thread_viewer/thread_viewer.tsx:43]: unused action

File: webapp/channels/src/packages/mattermost-redux/src/actions/posts.ts [hunk 3]
```diff
@@ -544,6 +544,16 @@ export function pinPost(postId: string) {
     };
 }
 
+/**
+ * Decrements the pinned post count for a channel by 1
+ */
+export function decrementPinnedPostCount(channelId: Channel['id']) {
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/packages/mattermost-redux/src/actions/posts.ts:550]: new action for decrementing pinned post in channel header

File: webapp/channels/src/packages/mattermost-redux/src/actions/posts.ts [hunk 4]
```diff
@@ -4,39 +4,39 @@
...
+import {GlobalState} from '@mattermost/types/store';
+import {Channel, ChannelUnread} from '@mattermost/types/channels';
+import {FetchPaginatedThreadOptions} from '@mattermost/types/client4';
+
 import {Client4, DEFAULT_LIMIT_AFTER, DEFAULT_LIMIT_BEFORE} from 'mattermost-redux/client';
 
 import {PostTypes, ChannelTypes, FileTypes, IntegrationTypes} from 'mattermost-redux/action_types';
 
+import {ActionResult, DispatchFunc, GetStateFunc} from 'mattermost-redux/types/actions';
+
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/packages/mattermost-redux/src/actions/posts.ts:19]: Unrelated change 🔄 : imports rearrangement

File: webapp/channels/src/packages/mattermost-redux/src/actions/posts.ts [hunk 5]
```diff
@@ -174,7 +174,7 @@ export function createPost(post: Post, files: any[] = []) {
         const pendingPostId = post.pending_post_id || `${currentUserId}:${timestamp}`;
         let actions: AnyAction[] = [];
 
-        if (Selectors.isPostIdSending(state, pendingPostId)) {
+        if (PostSelectors.isPostIdSending(state, pendingPostId)) {
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/packages/mattermost-redux/src/actions/posts.ts:177]: Unrelated change 🔄 : variable name change

File: webapp/channels/src/components/drafts/drafts.tsx [hunk 6]
```diff
@@ -77,7 +77,7 @@ function Drafts({
                         key={d.key}
                         displayName={displayName}
                         draft={d}
-                        isRemote={draftRemotes[d.key]}
+                        isRemote={draftRemotes?.[d.key]}
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/drafts/drafts.tsx:80]: Unrelated change 🔄 : This can be undefined

File: webapp/channels/src/components/post/index.tsx [hunk 7]
```diff
@@ -4,43 +4,41 @@
...
-
-import {ExtendedPost, removePost} from 'mattermost-redux/actions/posts';
-import {DispatchFunc, GetStateFunc} from 'mattermost-redux/types/actions';
-import {isThreadOpen} from 'selectors/views/threads';
-
-import {General} from 'mattermost-redux/constants';
-
+import {Locations, Preferences, RHSStates, StoragePrefixes} from 'utils/constants';
 import {getDisplayNameByUser} from 'utils/utils';
-import {getDirectTeammate} from 'mattermost-redux/selectors/entities/channels';
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/post/index.tsx:42]: Unrelated change 🔄 : imports rearrangment

File: webapp/channels/src/components/advanced_create_comment/advanced_create_comment.tsx [hunk 8]
```diff
@@ -350,7 +349,7 @@ class AdvancedCreateComment extends React.PureComponent<Props, State> {
     };
 
     saveDraftOnUnmount = () => {
-        if (!this.isDraftEdited || !this.state.draft) {
+        if (!this.isDraftEdited || !this.state.draft || this.props.rootDeleted) {
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/advanced_create_comment/advanced_create_comment.tsx:352]: Added the check here to not update draft on unmount

File: webapp/channels/src/components/post/actions.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+import {getGlobalItem} from 'selectors/storage';
+
+import {removeDraft} from 'actions/views/drafts';
+import {closeRightHandSide} from 'actions/views/rhs';
+
+import {StoragePrefixes} from 'utils/constants';
+
+import {GlobalState} from 'types/store';
+
+export function removePostCloseRHSDeleteDraft(post: ExtendedPost) {
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/post/actions.ts:17]: extracted and constructed on this function which also deletes the draft

File: webapp/channels/src/components/advanced_create_comment/advanced_create_comment.tsx (hunk 8 above)
Comment (hmhealey) [reply webapp/channels/src/components/advanced_create_comment/advanced_create_comment.tsx:352]: If I understand this right, it's to ensure that we don't remove the draft, close the RHS, and then immediately have this component re-save the draft. Is that right?

File: webapp/channels/src/components/post/post_options.tsx [hunk 10]
```diff
@@ -25,11 +25,10 @@ type Props = {
     post: Post;
     teamId: string;
     isFlagged: boolean;
-    removePost: (post: Post) => void;
+    removePostCloseRHSDeleteDraft: (post: Post) => void;
```
Comment (hmhealey) [inline webapp/channels/src/components/post/post_options.tsx:28]: 0/5 on changing it back, but I do like keeping the old name for actions like this. I don't think that the name of an action passed to the component needs to fully explain what it does as long as it's clear which button on the component it's triggered by.

File: webapp/channels/src/components/post/actions.ts (hunk 9 above)
Comment (hmhealey) [reply webapp/channels/src/components/post/actions.ts:21]: Yep, that makes sense to me. Having it in the index file is fine enough when there's not much going on in there, but there's already a lot happening in that file, so splitting it out makes sense to me

**Review by hmhealey — COMMENTED** [review]
This looks good to me, but I also wanted to check, does this remove the draft if you delete the post yourself? That skips showing the post with the X button to remove it

**Review by JulienTant — APPROVED** [review]

File: webapp/channels/src/components/advanced_create_comment/advanced_create_comment.tsx (hunk 8 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/advanced_create_comment/advanced_create_comment.tsx:352]: Yes, thats right.

File: webapp/channels/src/components/post/post_options.tsx (hunk 10 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/post/post_options.tsx:28]: yes, that makes sense. I changed it back

**Comment by M-ZubairAhmed** [comment]
> This looks good to me, but I also wanted to check, does this remove the draft if you delete the post yourself? That skips showing the post with the X button to remove it

The actions touched in this PR are not responsible for deleting the draft when you delete a post. That is done by `deleteAndRemovePost` action.

**Review by yasserfaraazkhan — APPROVED** [review]
thank you for the changes @M-ZubairAhmed

**Review by hmhealey — APPROVED** [review]
